### ISSUE TITLE
top+bottom margins ignored in scroll view

### ISSUE THREAD

**Issue by teleshoes** [buggins/coolreader#326]
top+bottom margin works in page-view-mode only. the values have no effect in scroll-view-mode.
left+right margins work in both page-view and scroll-view.

**Comment by teleshoes** [comment]
use cases:
1) i set bottom margin = 150px to not paint words underneath the android nav bar `<| o []` (pops up annoyingly on waydroid)
2) its also comfortable to hold my phone in such a way that my thumb covers the bottom 100px of my phone, especially in auto-scroll mode
3) separate issue, but this would ALSO provide a decent workaround for the problem of tts-readaloud toolbar covering the bottom 400px of the screen

**Comment by teleshoes** [comment]
i would also REALLY love to have the topbar with the clock+battery displayed in scroll-view-mode..

**Comment by teleshoes** [comment]
@plotn do you have any idea how this could be done? does knownreader by any chance have this feature?

**Comment by plotn** [comment]
@teleshoes, the scroll mode is seldom used (by community), as far as I know (but some people like). I even did (in KR) the ability "not to switch (internally) in scroll mode while TTS is active". Anyway, you can open issue in KR's repo - I will see if it could be done "not so expensive", but I dont promise the speed
(And I have plans to redesign TTS toolbar too, but it is also a big work (I want to make it similar to "common" toolbar).
Nowadays I am working to support external USB / SD card in modern androids, then to evolve offline dics (Lingvo DSL, StarDict, MDict), integrate Calibre Library and so on... )

**Comment by teleshoes** [comment]
ok, im trying that out right now, thanks! i assumed that TTS wouldnt work in page mode

**Comment by plotn** [comment]
I've caught your idea, but, be ready - it quite comlex "underneath" )

**Comment by teleshoes** [comment]
yea, i still see more words of the current selection in scroll-mode, even with the horrible tts toolbar covering 400px of my screen and the android nav-panel refusing to disappear in TTS mode.

my next idea is that i can tweak onSelectionCommand() to do SetPos() if the current selection enters the part of the screen that the toolbar covers

**Comment by teleshoes** [comment]
it SEEMS like it should be easy to add top+bottom margins to scroll-mode, but i havent been able to do it, and the ui-related code in crengine is scary..

**Comment by plotn** [comment]
Why ever you need to see the text, while you are listening to it?
In KR you can set transparent tts panel and transparent buttons onto it, thus almost everything become visible, maybe it is enough for you?

**Comment by teleshoes** [comment]
TTS mispronounces many words, especially foreign loan words and proper nouns, and it doesnt pause properly for parentheticals/etc, and the voice gets the wrong inflection for sentences when the usage may be ambiguous. i use TTS somewhere between a regular audiobook and an enhanced auto-scroll.

sometimes i am listening to TTS mostly like an audiobook, and only glance at the screen when the robot lady mangles a name (for example, while working or taking a walk).
other times, i am mostly reading along, and look away shortly when i need to (for example, to rinse the shampoo out of my hair).

either way, its useful to be able to alternate seamlessly between eyes and ears.


====
i set the bgcolor of the TTS toolbar and all buttons to alpha=0 in coolreader already, but the sliders and buttons still block a word or two. also, now the words make it hard to see the toolbar sliders.
in my opinion, it is simply bad design to have a transparent/translucent toolbar overlay over words. the alpha channel just makes both UIs worse.

=> the TTS toolbar should be MUCH smaller, with an optional additional toolbar with volume+speed, and it should be opaque, and it should subtract from the screen height.

better yet, the toolbar could have a button to hide it completely. tapping anywhere, instead of ending TTS, would pause TTS, and show the toolbar.

**Comment by plotn** [comment]
I finally have such idea:
1. Clicking outside TTS toolbar will not stop reading;
2. While reading the first two buttons on the toolbar (if visible) and in the menu (usually center tap) will become "show reading toolbar", "stop tts reading".
I think this will be enough.

**Comment by teleshoes** [comment]
if i understood you, that sounds strictly worse than simply showing the toolbar on tap-anywhere

A) if you want to stop (or pause), you STILL need to click twice
B) if you want to change volume/etc, you need to click THREE times instead of only twice

there is no workflow that will have FEWER clicks in your model.

**Comment by plotn** [comment]
Okay, need more time to think

**Comment by plotn** [comment]
@teleshoes , see https://4pda.to/forum/index.php?s=&showtopic=995536&view=findpost&p=113718947 
i think the result is what you wanted...

**Comment by teleshoes** [comment]
yes, you describe exactly what i want, i think! 404 on the apk link tho so i cant test it. any chance you could push it to a branch so i could compile it?

**Comment by plotn** [comment]
@teleshoes, register at this forum, so 404 will disappear.
Sure, i can push the changes - maybe in the evening.

**Comment by plotn** [comment]
@teleshoes , see two latest commits:
https://github.com/plotn/coolreader/commit/78f29634ae24ff2d4a2b49565c5597b2a1ac1940

**Comment by teleshoes** [comment]
yea, its perfect! love the top-line vs middle behavior.
im gonna tweak it a bit (e.g.: ignore isSpeaking and always toggle ui on middle-click) and pull it into my coolreader fork.
thanks so much for doing this!

**Comment by plotn** [comment]
@teleshoes , hi! not about this topic, but maybe it would be interesting... see new features: 
https://4pda.to/forum/index.php?s=&showtopic=995536&view=findpost&p=113801152
